CRM-194 Add in new template for report to generate recurring contribu…

seamuslee001 opened the pull request:
…tion summary

@joeMurray I believe this will achieve the job but would appreciate someone doing some review first

monishdeb commented:
I have deployed this patch in test2-dms and created a report instance including all the necessary columns needed. Following is how the new report look [Recurring Contribution (Detailed)](https://test2-dms.canadahelps.org/civicrm/report/instance/50?reset=1&force=1) look : 
<img width="1290" alt="Screen Shot 2020-03-04 at 1 02 21 PM" src="https://user-images.githubusercontent.com/3735621/75855664-c07f0f80-5e18-11ea-8aa5-d6e58ebd0926.png">
 
UI changes/fixes needed: 
1. As you can see in the screenshot it throw a error message 'Unknown extension key in a translation string: chreports'. Can you please fix it?
2. Please adjust the order by placing 'Contact ID' before 'Contact Name' column as per the [sheet Tab 21](https://docs.google.com/spreadsheets/d/1_gIfvhBHNEnjNPyEUnhfQoeBUBGROelWvlXgkMAvT5Y/edit#gid=236368217) 
3. Rename  'Street Address' to 'Address - Primary'
4. Phone display column is missing, please add it.
5. Place 'Email' column after 'Phone'. Currently, it is shown at last.
6. Column 'Last Month Amount' is missing the currency symbol. (add ```type``` as ```CRM_Utils_TYPE::T_MONEY```)
7. Statistics section is missing. Although it is not mentioned in the doc but do we need to add infomation like 'Total Contacts', 'Total Recurring Contributions', 'Total Amount' etc. ? 

Technical changes: 
1. The SQL query is taking long time to fetch result. May be we should move the OR condition in the WHERE clause to JOINs by using a derived table?
2. Can we replace the hardcoded custom field column and custom group table name by using util function which fetch custom table and custom column name? 
Like replacing 
```
  LEFT JOIN civicrm_value_contribution__15 csummary ON csummary.entity_id = {$this->_aliases['civicrm_contribution']}.id
```
with 
```php
$tableName = Utils::_getTableNameByName('Contribution_Details');
...
  LEFT JOIN {$tableName} csummary ON csummary.entity_id = {$this->_aliases['civicrm_contribution']}.id
```
3. Can you add the mgd file for created report instance under https://github.com/CanadaHelps/chreports/tree/master/managed 
4. Please change the template name [here](https://github.com/CanadaHelps/chreports/pull/13/files#diff-151e6acbef623ad247f219406f247869R15) to 'Recurring Contribution (Detailed)' 
I have left some comment in the patch itself. Please have a look.

monishdeb commented on CRM/Chreports/Form/Report/RecurSummary.php:
```diff
@@ -0,0 +1,248 @@
+    $this->_whereClauses[] = '( contribution_civireport.contribution_recur_id <> 0 OR cd.sg_flag_38 <> 0 )';
+    $this->_whereClauses[] = '( contribution_last_month.contribution_recur_id <> 0 OR csummary.sg_flag_38 <> 0 )';
+    $this->_whereClauses[] = '( contribution_summary.contribution_recur_id <> 0 OR clastMonth.sg_flag_38 <> 0 )';
+    if (empty($this->_whereClauses)) {
```
I think this condition is not needed because $this->_whereClauses is not empty as it is prefilled with clauses from 169-172

monishdeb commented on CRM/Chreports/Form/Report/RecurSummary.php:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  function preProcess() {
+    $this->assign('reportTitle', E::ts('Contribution Recur Summary'));
```
I think it is not required as it will override the title of report instances

JoeMurray replied:
The Report Description tab in the spreadsheet has 'Recurring Contributions (Summary)'. I'm not sure what is authoritative, but your name seems more like a CiviCRM terminology than a CH one.

monishdeb commented on CRM/Chreports/Form/Report/RecurSummary.mgd.php:
```diff
@@ -0,0 +1,22 @@
+    array (
+      'version' => 3,
+      'label' => 'RecurSummary',
+      'description' => 'RecurSummary (biz.jmaconsulting.chreports)',
```
As per Tab1 titled 'Report Descriptions', it should be  'Overview of Recurring Contributions'

JoeMurray commented on CRM/Chreports/Form/Report/RecurSummary.php:
```diff
@@ -0,0 +1,248 @@
+               LEFT JOIN civicrm_value_contribution__15 cd ON cd.entity_id = {$this->_aliases['civicrm_contribution']}.id
+               INNER JOIN civicrm_contribution contribution_summary ON contribution_summary.contact_id = {$this->_aliases['civicrm_contribution']}.contact_id
+                 AND contribution_summary.is_test = 0
+               LEFT JOIN civicrm_value_contribution__15 csummary ON csummary.entity_id = {$this->_aliases['civicrm_contribution']}.id
```
I'm having trouble seeing why we need to left join civicrm_value_contribution__15 twice. Might make sense, but I'm not sure.

seamuslee001 replied:
@joeMurray my thinking was that we would be limiting the contributions in the first join to those that happened this month the 2nd set of joins would be for all time and the 3rd set of joins would be for the previous month. So I figure each time we join onto essentially the civicrm_contribution table we would need to create a new join to civicrm_value_contribution__15

JoeMurray commented on CRM/Chreports/Form/Report/RecurSummary.php:
```diff
@@ -0,0 +1,248 @@
+    $this->_whereClauses[] = $this->dateClause('contribution_last_month.receive_date', 'previous.month', NULL, NULL, CRM_Utils_Type::T_DATE + CRM_Utils_Type::T_TIME, NULL, NULL);
+    $this->_whereClauses[] = '( contribution_civireport.contribution_recur_id <> 0 OR cd.sg_flag_38 <> 0 )';
+    $this->_whereClauses[] = '( contribution_last_month.contribution_recur_id <> 0 OR csummary.sg_flag_38 <> 0 )';
+    $this->_whereClauses[] = '( contribution_summary.contribution_recur_id <> 0 OR clastMonth.sg_flag_38 <> 0 )';
```
I think I need to see the whole query to understand why it makes sense to test the same two fields three times, and have those tables joined three separate times.

monishdeb replied:
Please use the created report instance's https://test2-dms.canadahelps.org/civicrm/report/instance/50?reset=1&force=1 Developer tab for the SQL

JoeMurray commented:
The spec shows two statistics fields only. Please add 'Last Month Amount Total' and 'This Month Amount Total'.
I've rejigged the query to remove several joins that would be exploding the number of rows returned, got rid of a select field or two eg exposed contact, changed sort_name to display_name (please confirm against what is used for other reports) and added other select fields, added a group by on the source column so it would display valid values, and added an order by:

```
SELECT contact_civireport.id as civicrm_contact_id, 
contact_civireport.display_name as civicrm_contact_display_name, 
address_civireport.street_address as civicrm_address_street_address,
phone_civireport.phone as civicrm_phone_phone,
email_civireport.email as civicrm_email_email,
contribution_civireport.source as civicrm_contribution_source, 
SUM(IF(contribution_civireport.receive_date>=20200201000000 AND contribution_civireport.receive_date<=20200229235959,contribution_civireport.total_amount,0)) as civicrm_contribution_last_month_amount, 
SUM(IF(contribution_civireport.receive_date>=20200301000000 AND contribution_civireport.receive_date<=20200331235959,contribution_civireport.total_amount,0)) as civicrm_contribution_this_month_amount,
MIN(contribution_civireport.receive_date) as civicrm_contribution_start_date, 
COUNT(DISTINCT contribution_civireport.id) as civicrm_contribution_completed_contributions

FROM  civicrm_contact contact_civireport                 
INNER JOIN civicrm_contribution contribution_civireport ON contact_civireport.id = contribution_civireport.contact_id AND contribution_civireport.is_test = 0
LEFT JOIN civicrm_value_contribution__15 cd ON cd.entity_id = contribution_civireport.id                
LEFT JOIN civicrm_address address_civireport ON (contact_civireport.id = address_civireport.contact_id) AND address_civireport.is_primary = 1
LEFT JOIN civicrm_email email_civireport ON contact_civireport.id = email_civireport.contact_id AND email_civireport.is_primary = 1
LEFT JOIN civicrm_phone phone_civireport ON contact_civireport.id = phone_civireport.contact_id AND phone_civireport.is_primary = 1

WHERE ( contribution_civireport.contribution_recur_id <> 0 OR cd.sg_flag_38 <> 0 ) 

GROUP BY contact_civireport.id, contribution_civireport.source
ORDER BY contact_civireport.sort_name
LIMIT 0, 50;


```

monishdeb commented:
@seamuslee001 @JoeMurray 

I have changed query as per above [comment](https://github.com/CanadaHelps/chreports/pull/13#issuecomment-594770217) and also fixed few issues, which I have strikedout in the PR description.

 
Prepared a new report instance after the changes
https://test2-dms.canadahelps.org/civicrm/report/instance/53?reset=1&force=1

The report will show no result because filter recieve_date is set to 'Current month' thus there are no recurring contributions in this month. Please change the date filter to find more result.

seamuslee001 commented:
@monishdeb @JoeMurray I have now updated the report to include the $ totals as per the spreadsheet and updated the description as per Monish's comment. I tested the report and the report seems to work correctly on test2. Joe are you able to review the query as well.

monishdeb commented:
Changes look great. I can see the Grand total row at the end. 
<img width="1286" alt="Screen Shot 2020-03-06 at 11 20 45 AM" src="https://user-images.githubusercontent.com/3735621/76056921-d538e000-5f9d-11ea-9395-5daec82784e9.png">

I found a core bug where the grand total always shows in the last page, meaning if there are 100 rows distributes in 2 pages(2x50). The Grand total row only shows up in the last/2nd page.
